### ISSUE TITLE
Design Patterns concepts implementations in Python

### ISSUE THREAD

**Issue by rhoitjadhav** [trekhleb/learn-python#35]
It would be very nice if you could implement design patterns in Python. I hope you are aware of these concepts.

For more info please navigate to this [link](https://en.wikipedia.org/wiki/Software_design_pattern)

**Comment by evaldez89** [comment]
I agreed. People often discard python as an OOP language and use it only for its scripting advantages. Nonetheless, it it as powerful as any other OOP typed language.

Here, I started a project implementing abstract factory. It needs improvement, but it's a start.

https://github.com/evaldez89/banks-transaction-email-reader

**Comment by Sidibeluis** [comment]
Y

**Comment by sveneschlbeck** [comment]
@rhoitjadhav @evaldez89 I agree, too. Maybe, let's start by building up a repo folder structure and then fill it over time with good examples. But imho, structural and behavioral patterns should be covered at least...

**Comment by sveneschlbeck** [comment]
@rhoitjadhav @evaldez89 @giugiozza @ANONYJINOKA @sebaslucio-actuary @trekhleb I added a first structural frame (directory backbone) with **some first examples** that I found really useful. Since many design principles are already partially or fully integrated into today's core-python, there are not THAT many design patterns out there. Nevertheless the first examples should give a nice overview and make devs appreciate python even more.  
Feel free to add more example classes/functions to the specific directories/modules.
